python/ray/serve/api.py

```diff
@@ -643,6 +643,16 @@ def run(
 
     client.deploy_group(parameter_group, _blocking=_blocking)
 
+    # clean up the old deployments
+    new_deployments_names = set()
+    for deployment in deployments:
+        new_deployments_names.add(deployment._name)
+
+    all_deployments = list_deployments()
+    all_deployments_names = set(all_deployments.keys())
+    deployment_names_to_delete = all_deployments_names.difference(new_deployments_names)
+    client.delete_deployments(deployment_names_to_delete)
+
```

Comment:
should this just be put into `deploy_group` instead of doing it client-side? that'd accomplish the same thing but probably be more robust. @shrekris-anyscale WDYT?

Reply:
I think @shrekris-anyscale made good point for the serve.run() restriction. It might not be ideal to add it.

@shrekris-anyscale do you think we can directly use client api instead of using serve.run() in the ServeHead? In this case, we can move the clean up logics into the client side, and use one async call to do the deploy and clean up. if yes, I plan to add a flag into the deploy_group to indicate whether we remove the current deployment or not.

Reply:
I think this is probably an OK restriction to be honest, but for now the safe thing would be to introduce a flag as @sihanwang41 suggests and we can set it to `true` in the API server and `false` in interactive dev?

Reply:
Actually, after some more thought, I think it's okay to make `serve.run()` delete all other deployments. Ideally, `serve.run()` should be idempotent, but it's impossible to distinguish between outdated deployments and deployments that belong to separate Deployment Graphs. For idempotency, we can make `serve.run()` tear down all the deployments not in the passed-in node or app.

We should still use only `serve.run()` inside `serve_head.py`. However, we should move the combined deploy and deletion logic into the client and the controller. Here are the changes we should make:

* The client should offer a new function:

```python
# In client.py

def replace_deployments(self, deployments: List[Dict], remove_names: Optional[List[String]]=None, _blocking: bool=True):
"""Atomically replaces running deployments in Ray cluster with passed-in deployments.

Args:
    deployments [List[Dict]]: A list of all the deployment parameters. Same as "deployments" argument from
        deploy_group() function.
    remove_names Optional[List[String]]: A list of names for deployments to remove. If None, all deployments
        not specified in the deployments argument will be removed. Same as "names" argument from
        delete_deployments() function.
"""
    ...
    self._controller.replace_deployments()
    ...
```

* The `ServeController` needs to offer a single synchronous function `replace_deployments()`. If the client makes separate calls to the controller's `deploy_group()` and `delete_deployments()` functions, then the controller could execute asynchronous functions between the calls, which means the replace action would not be atomic. The client can then call the `ServeController`'s new `replace_deployments()` function.

* Then `serve.run()` could call the client's `replace_deployments()` function with `remove_names=None` to atomically replace deployments.

Reply:
Hi @shrekris-anyscale ,  I think there is issue I'd like to make a consensus here.
Do we want to make the replace_group as first citizen and directly remove the deploy_group (serve.run is the only place to use the deploy_group from my search)? If I implement the replace_group on the client.py side, I don't think deploy_group need to be existed in the code. That is why I think we can add a flag in deploy_group for now to control whether we remove the deployments or not. WDYT?

I can make the adjustments on the controller side as you suggested.

Reply:
Good question– I think ideally we should have only `replace_deployments()`, and we should remove `deploy_group()` and `delete_deployments()`. However, since that's a somewhat large change, I'm okay with introducing a flag inside `deploy_group()` to remove deployments.